mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,182 @@
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+import itertools
+
+from joblib import Parallel, delayed
+import numpy as np
+
+
+def _reconstruct_case(grid_search_test_case, key_names, init_classes,
+                      kspace_data, fourier_params):
+    """Internal Function to carry out reconstruction for a
+    special case. This function pulls in appropriate keyword arguments
+    from input and declares appropriate Linear, Fourier and Regularizer
+    Operators. These operators later are used to create the image model
+    by defining the reconstructor. Then the reconstruction is carried and
+    results are returned
+
+    Parameters
+    ----------
+    grid_search_test_case: list of parameters
+        keyword arguments for different operators for reconstruction
+    key_names: list of strings
+        names of keyword arguments specified for reconstruction in 'case'
+    init_classes: list of classes
+        classes for varoius operators. must be in format
+        [linear, regularizer, reconstructor]
+    kspace_data: np.ndarray
+        the kspace data for reconstruction
+    fourier_params: dict
+        holds dictionary with init_class pointing to fourier class to be used
+        and args having keyword arguments for initialization
+        NOTE: We declare fourier operator inside this function to allow
+        parallel execution as NonCartesianFFT cannot be pickled.
+    """
+    fourier_op = fourier_params['init_class'](**fourier_params['kwargs'])
+    linear_op = init_classes[0](
+        **dict(zip(key_names[0], grid_search_test_case[0])))
+    regularizer_op = init_classes[1](
+        **dict(zip(key_names[1], grid_search_test_case[1]))
+    )
+    reconstructor = init_classes[2](
+        fourier_op=fourier_op,
+        linear_op=linear_op,
+        regularizer_op=regularizer_op,
+        **dict(zip(key_names[2], grid_search_test_case[2]))
+    )
+    raw_results = reconstructor.reconstruct(
+        kspace_data=kspace_data,
+        **dict(zip(key_names[3], grid_search_test_case[3])),
+    )
+    return raw_results
+
+
+def gather_result(metric, metric_direction, results):
+    """ Gather the best reconstruction result.
+
+    Parameters:
+    -----------
+    metric: str,
+        the name of the metric, it will become a dict key in the output dict.
+    metric_direction: bool,
+        if True the higher the better the metric value is (like for `ssim`),
+        else the lower the better.
+    results: list of list,
+        list of the raw results of the gridsearch
+
+    Return:
+    -------
+    results and location of best results in given set of raw results
+    """
+    list_metric = np.array([res[2][metric]['values'][-1]
```

Comment:
nitpicking, but always try to make sure the multi-line statements are not going to cause potential unnecessary changes in the future.

Reply:
Fixed

Reply:
here I would argue the most readable way is:
```python
list_metric = np.array([
	res[2][metric]['values'][-1] for res in results
])
```